segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java

```diff
@@ -0,0 +1,857 @@
+     * @throws IOException            If the stream threw one.
+     * @throws SerializationException If the given Stream is an invalid metadata serialization.
+     * @throws IllegalStateException  If the Metadata is not in Recovery Mode.
+     */
+    private void deserializeFrom(MetadataCheckpointOperation operation) throws IOException, SerializationException {
+        Preconditions.checkState(this.recoveryMode, "Cannot deserialize Metadata in non-recovery mode.");
+
+        DataInputStream stream = new DataInputStream(new GZIPInputStream(operation.getContents().getReader()));
+
+        // 1. Version.
+        byte version = stream.readByte();
+        if (version != CURRENT_SERIALIZATION_VERSION) {
+            throw new SerializationException("Metadata.deserialize", String.format("Unsupported version: %d.", version));
+        }
+
+        // 2. Container id.
+        int containerId = stream.readInt();
+        if (this.containerId != containerId) {
+            throw new SerializationException("Metadata.deserialize",
+                    String.format("Invalid ContainerId. Expected '%d', actual '%d'.", this.containerId, containerId));
+        }
+
+        // This is not retrieved from serialization, but rather from the operation itself.
+        setOperationSequenceNumber(operation.getSequenceNumber());
+
+        // 3. Stream Segments (unchanged).
+        int segmentCount = stream.readInt();
+        for (int i = 0; i < segmentCount; i++) {
+            deserializeSegmentMetadata(stream);
+        }
+
+        // 4. Stream Segments (updated).
+        segmentCount = stream.readInt();
+        for (int i = 0; i < segmentCount; i++) {
+            deserializeSegmentMetadata(stream);
+        }
+
+        // 5. New Stream Segments.
+        segmentCount = stream.readInt();
+        for (int i = 0; i < segmentCount; i++) {
+            deserializeSegmentMetadata(stream);
+        }
+    }
+
+    /**
+     * Applies the updates stored in the given StorageMetadataCheckpointOperation.
+     *
+     * @param operation The StorageMetadataCheckpointOperation to update from.
+     * @throws IOException            If the stream threw one.
+     * @throws SerializationException If the given Stream is an invalid metadata serialization.
+     * @throws IllegalStateException  If the Metadata is not in Recovery Mode.
+     */
+    private void updateFrom(StorageMetadataCheckpointOperation operation) throws IOException, SerializationException, MetadataUpdateException {
+        Preconditions.checkState(this.recoveryMode, "Cannot bulk-update Metadata in non-recovery mode.");
+
+        DataInputStream stream = new DataInputStream(new GZIPInputStream(operation.getContents().getReader()));
+
+        // 1. Version.
+        byte version = stream.readByte();
+        if (version != CURRENT_SERIALIZATION_VERSION) {
+            throw new SerializationException("Metadata.updateFrom", String.format("Unsupported version: %d.", version));
```

Comment:
As of right now we only have one Version. As we have multiple versions, this code will be updated to handle that and be backwards compatible. This check just verifies that we don't have an unsupported version - I expect that to stay there in addition to various version handling.